server/src/main/java/org/elasticsearch/cluster/metadata/MetadataUpdateSettingsService.java

```diff
@@ -87,183 +98,218 @@ public MetadataUpdateSettingsService(
 
-                return updatedState;
+            try {
+                for (Index index : openIndices) {
+                    final IndexMetadata currentMetadata = currentState.metadata().getIndexSafe(index);
+                    final IndexMetadata updatedMetadata = updatedState.metadata().getIndexSafe(index);
+                    indicesService.verifyIndexMetadata(currentMetadata, updatedMetadata);
+                }
+                for (Index index : closedIndices) {
+                    final IndexMetadata currentMetadata = currentState.metadata().getIndexSafe(index);
+                    final IndexMetadata updatedMetadata = updatedState.metadata().getIndexSafe(index);
+                    // Verifies that the current index settings can be updated with the updated dynamic settings.
+                    indicesService.verifyIndexMetadata(currentMetadata, updatedMetadata);
+                    // Now check that we can create the index with the updated settings (dynamic and non-dynamic).
+                    // This step is mandatory since we allow to update non-dynamic settings on closed indices.
+                    indicesService.verifyIndexMetadata(updatedMetadata, updatedMetadata);
+                }
+            } catch (IOException ex) {
+                throw ExceptionsHelper.convertToElastic(ex);
             }
-        };
 
-        clusterService.submitStateUpdateTask("update-settings " + Arrays.toString(request.indices()), clusterTask, this.executor);
+            return updatedState;
+        }
+    }
+
+    record MetadataUpdateSettingsTask(UpdateSettingsClusterStateUpdateRequest request, ActionListener<AcknowledgedResponse> listener)
+        implements
+            ClusterStateTaskListener,
+            ClusterStateAckListener {
+
+        @Override
+        public void onFailure(Exception e) {
+            listener.onFailure(e);
+        }
+
+        @Override
+        public boolean mustAck(DiscoveryNode discoveryNode) {
+            return true;
+        }
+
+        @Override
+        public void onAllNodesAcked() {
+            listener.onResponse(AcknowledgedResponse.of(true));
+        }
+
+        @Override
+        public void onAckFailure(Exception e) {
+            listener.onResponse(AcknowledgedResponse.of(false));
+        }
+
+        @Override
+        public void onAckTimeout() {
+            listener.onResponse(AcknowledgedResponse.FALSE);
+        }
+
+        @Override
+        public TimeValue ackTimeout() {
+            return request.ackTimeout();
+        }
+
+        @Override
+        public void clusterStateProcessed(ClusterState oldState, ClusterState newState) {
+            assert false : "not called";
+        }
+    }
+
+    public void updateSettings(final UpdateSettingsClusterStateUpdateRequest request, final ActionListener<AcknowledgedResponse> listener) {
+        ActionListener<AcknowledgedResponse> actionListener = wrapPreservingContext(listener, threadPool.getThreadContext());
```

Comment:
Hmm I think this wrapping is very old and hasn't been necessary for ages. Could you add an `ESSingleNodeTestCase` test which sets up a thread context, calls the update settings API, and asserts that the context is passed through to the listener (even without this wrapping). See e.g. how `MasterServiceTests#testTaskFailureNotification` interacts with the thread context.